pkg/virt-launcher/virtblocks/notify.go

```diff
@@ -0,0 +1,286 @@
+	response, err := n.v1client.HandleDomainEvent(ctx, &request)
+
+	if err != nil {
+		return err
+	} else if response.Success != true {
+		msg := fmt.Sprintf("failed to notify domain event: %s", response.Message)
+		return fmt.Errorf(msg)
+	}
+
+	return nil
+}
+
+func newWatchEventError(err error) watch.Event {
+	return watch.Event{Type: watch.Error, Object: &metav1.Status{Status: metav1.StatusFailure, Message: err.Error()}}
+}
+
+func eventCallback(c *VirtBlocks, domain *api.Domain, client *NotifierImpl, events chan watch.Event, interfaceStatus *[]api.InterfaceStatus) {
+	d, err := c.GetDomain()
+	if err != nil {
+		if !isNotFound(err) {
+			log.Log.Reason(err).Error("Could not fetch the Domain.")
+			client.SendDomainEvent(newWatchEventError(err))
+			return
+		}
+		domain.SetState(api.NoState, api.ReasonNonExistent)
+	} else {
+		// No matter which event, try to fetch the domain xml
+		// and the state. If we get a IsNotFound error, that
+		// means that the VirtualMachineInstance was removed.
+		status, reason, err := d.GetState()
+		if err != nil {
+			if !isNotFound(err) {
+				log.Log.Reason(err).Error("Could not fetch the Domain state.")
+				client.SendDomainEvent(newWatchEventError(err))
+				return
+			}
+			domain.SetState(api.NoState, api.ReasonNonExistent)
+		} else {
+			domain.SetState(status, reason)
+		}
+
+		// XXX metadata is missing here
+		domain, err := d.Spec()
+		if err != nil {
+			// NOTE: Getting domain metadata for a live-migrating VM isn't allowed
+			if !isNotFound(err) {
+				log.Log.Reason(err).Error("Could not fetch the Domain specification.")
+				client.SendDomainEvent(newWatchEventError(err))
+				return
+			}
+		} else {
+			domain.ObjectMeta.UID = domain.Spec.Metadata.KubeVirt.UID
+		}
+		log.Log.Infof("kubevirt domain status: %v(%v):%v(%v)", domain.Status.Status, status, domain.Status.Reason, reason)
+	}
+
+	switch domain.Status.Reason {
+	case api.ReasonNonExistent:
+		watchEvent := watch.Event{Type: watch.Deleted, Object: domain}
+		client.SendDomainEvent(watchEvent)
+		events <- watchEvent
+	default:
+		if interfaceStatus != nil {
+			domain.Status.Interfaces = *interfaceStatus
+			event := watch.Event{Type: watch.Modified, Object: domain}
+			client.SendDomainEvent(event)
+			events <- event
+		}
+		client.SendDomainEvent(watch.Event{Type: watch.Modified, Object: domain})
+	}
+}
+
+func (n *NotifierImpl) StartDomainNotifier(domainConn *VirtBlocks, deleteNotificationSent chan watch.Event, vmiUID types.UID, qemuAgentPollerInterval *time.Duration) error {
+	eventChan := make(chan libvirtEvent, 10)
+
+	// Run the event process logic in a separate go-routine to not block libvirt
+	go func() {
+		var interfaceStatuses *[]api.InterfaceStatus
+		for {
+			select {
+			case event := <-eventChan:
+				domain := util.NewDomainFromName(event.Domain, vmiUID)
+				eventCallback(domainConn, domain, n, deleteNotificationSent, interfaceStatuses)
+				log.Log.Info("processed event")
+			}
+		}
+	}()
+
+	go func() {
```

Comment:
Instead of listening to libvirt events, this right now simply polls every second "something" and checks if something interesting happened.

Reply:
Next I will update the tests. Then a full shell for virtblocks is ready which does not depend on anything from libvirt.